File: terrain_generation/terrain_generator.py
import numpy as np
import noise

class TerrainGenerator:
    def __init__(self, width, height, seed=None):
        self.width = width
        self.height = height
        self.seed = seed if seed is not None else np.random.randint(0, 1000)

    def generate_terrain(self, scale=50.0, octaves=6, persistence=0.5, lacunarity=2.0):
        terrain = np.zeros((self.width, self.height))
        for x in range(self.width):
            for y in range(self.height):
                nx = x / self.width - 0.5
                ny = y / self.height - 0.5
                value = noise.snoise2(
                    nx * scale,
                    ny * scale,
                    octaves=octaves,
                    persistence=persistence,
                    lacunarity=lacunarity,
                    repeatx=self.width,
                    repeaty=self.height,
                    base=self.seed,
                )
                terrain[x][y] = value
        return terrain

File: terrain_generation/terrain_generator.cpp
#include <iostream>
#include <vector>
#include <random>

// Генерирует случайный ландшафт в 3D
std::vector<std::vector<std::vector<int>>> GenerateTerrain(int width, int length, int height) {
    std::vector<std::vector<std::vector<int>>> terrain;

    // Инициализируем генератор случайных чисел
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 255);

    // Создаем 3D массив для хранения высот точек ландшафта
    for (int x = 0; x < width; ++x) {
        std::vector<std::vector<int>> row;
        for (int y = 0; y < length; ++y) {
            std::vector<int> column;
            for (int z = 0; z < height; ++z) {
                // Генерируем случайную высоту от 0 до 255
                int elevation = dis(gen);
                column.push_back(elevation);
            }
            row.push_back(column);
        }
        terrain.push_back(row);
    }

    return terrain;
}

int main() {
    int width = 100;
    int length = 100;
    int height = 50;

    std::vector<std::vector<std::vector<int>>> terrain = GenerateTerrain(width, length, height);

    // Выводим сгенерированный ландшафт
    for (int x = 0; x < width; ++x) {
        for (int y = 0; y < length; ++y) {
            for (int z = 0; z < height; ++z) {
                std::cout << "Terrain at (" << x << ", " << y << ", " << z << "): " << terrain[x][y][z] << std::endl;
            }
        }
    }

    return 0;
}
